pallets/parachain-staking/src/migrations.rs

```diff
@@ -17,19 +17,117 @@
 //! # Migrations
 use crate::{
 	pallet::{migrate_nominator_to_delegator_state, RoundIndex},
-	BalanceOf, CandidateState, CollatorCandidate, CollatorState2, Config, DelegatorState,
-	ExitQueue2, NominatorState2, Points, Round, Staked,
+	BalanceOf, Bond, CandidateState, CollatorCandidate, CollatorState2, Config, DelegatorState,
+	ExitQueue2, NominatorState2, Pallet, Points, Round, Staked,
 };
 #[cfg(feature = "try-runtime")]
 use crate::{Collator2, Delegator, Nominator2};
 #[cfg(feature = "try-runtime")]
+use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+#[cfg(feature = "try-runtime")]
 use frame_support::Twox64Concat;
+extern crate alloc;
+#[cfg(feature = "try-runtime")]
+use alloc::format;
 use frame_support::{
 	pallet_prelude::PhantomData,
 	traits::{Get, OnRuntimeUpgrade},
 	weights::Weight,
 };
-use sp_std::collections::btree_map::BTreeMap;
+use sp_std::{collections::btree_map::BTreeMap, vec::Vec};
+
+/// Migration to properly increase maximum delegations per collator
+/// This migration can be used to recompute the top and bottom delegations whenever
+/// MaxDelegatorsPerCandidate changes (works for decrease as well)
+pub struct IncreaseMaxDelegationsPerCandidate<T>(PhantomData<T>);
+impl<T: Config> OnRuntimeUpgrade for IncreaseMaxDelegationsPerCandidate<T> {
+	fn on_runtime_upgrade() -> Weight {
+		let (mut reads, mut writes) = (0u64, 0u64);
+		for (account, state) in <CandidateState<T>>::iter() {
+			reads += 1u64;
+			// 1. collect all delegations into single vec and order them
+			let mut all_delegations = state.top_delegations.clone();
+			let mut starting_bottom_delegations = state.bottom_delegations.clone();
+			all_delegations.append(&mut starting_bottom_delegations);
+			// sort all delegations from greatest to least
+			all_delegations.sort_unstable_by(|a, b| b.amount.cmp(&a.amount));
+			let top_n = T::MaxDelegatorsPerCandidate::get() as usize;
+			// 2. split them into top and bottom using the T::MaxNominatorsPerCollator
+			let top_delegations: Vec<Bond<T::AccountId, BalanceOf<T>>> =
+				all_delegations.iter().take(top_n).cloned().collect();
+			let bottom_delegations = if all_delegations.len() > top_n {
+				let rest = all_delegations.len() - top_n;
+				let bottom: Vec<Bond<T::AccountId, BalanceOf<T>>> = all_delegations.iter().rev().take(rest).cloned().collect();
+					.clone()
```

Comment:
Syntax error, seems code has been duplicated while formatting ?

Reply:
yes, I pushed the fix. I was just applying the suggestion without looking